inspektor-gadget/inspektor-gadget
# Add support for TC programs

This PR implements support for TC programs in inspector gadget. 

The main component of this PR is a new `tchandler` package that handles how TC programs are attached to containers and network interfaces. (very similar to the existing `pkg/gadgets/internal/networktracer` one). Programs are attached by using a clsact qdisc and filter in direct-action mode (check https://qmonnet.github.io/whirl-offload/2020/04/11/tc-bpf-direct-action/ to learn more).

For containers, the programs are attached to the peer interface on the host, this guarantees that processes running inside the container can't detach them from the interface. We might considering adding support for attaching these in the container side later on.

Gadgets with multiple TC programs, multiple gadgets and multiple instances of Inspektor Gadget running at the same time are supported. This is done by using a different handle each time a program is attached to the qdisc.

### How to test

Define a TC program, for instance, this one is a very simple firewall: (It's not supported to fill the ips map yet, to this only works to print messages to trace_pipe)

```c
#include <stdint.h>
#include <linux/bpf.h>
#include <bpf/bpf_helpers.h>
#include <linux/if_ether.h>
#include <linux/ip.h>
#include <linux/swab.h>
#include <linux/pkt_cls.h>
#include <bpf/bpf_endian.h>

char __license[] SEC("license") = "GPL";

struct {
	__uint(type, BPF_MAP_TYPE_HASH);
	__uint(max_entries, 1024);
	__type(key, __u32); // ipv4
	__type(value, __u8); // not used
} ips SEC(".maps");

SEC("classifier/egress/drop")
int egress_drop(struct __sk_buff *skb) {
	//bpf_printk("Packet on egress");

	void *data_end = (void *)(unsigned long long)skb->data_end;
	void *data = (void *)(unsigned long long)skb->data;

	// Check if the packet is not malformed
	struct ethhdr *eth = data;
	if (data + sizeof(struct ethhdr) > data_end)
		return TC_ACT_SHOT;

	// Check that this is an IP packet
	if (bpf_ntohs(eth->h_proto) != ETH_P_IP)
		return TC_ACT_UNSPEC;

	// Check if the packet is not malformed
	struct iphdr *ip = data + sizeof(struct ethhdr);
	if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) > data_end)
		return TC_ACT_SHOT;

	bpf_printk("egress: src:%pi4 -> dst: %pi4", &ip->saddr, &ip->daddr);

	// Drop packet if in map
	__u8 *dummy = bpf_map_lookup_elem(&ips, &ip->daddr);
	if (dummy) {
		bpf_printk("Dropping in egress");
		return TC_ACT_SHOT;
	}

	return TC_ACT_UNSPEC;
}

SEC("classifier/ingress/drop")
int ingress_drop(struct __sk_buff *skb) {
	//bpf_printk("Packet on ingress");

	void *data_end = (void *)(unsigned long long)skb->data_end;
	void *data = (void *)(unsigned long long)skb->data;

	// Check if the packet is not malformed
	struct ethhdr *eth = data;
	if (data + sizeof(struct ethhdr) > data_end)
		return TC_ACT_SHOT;

	// Check that this is an IP packet
	if (bpf_ntohs(eth->h_proto) != ETH_P_IP)
		return TC_ACT_UNSPEC;

	// Check if the packet is not malformed
	struct iphdr *ip = data + sizeof(struct ethhdr);
	if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) > data_end)
		return TC_ACT_SHOT;

	bpf_printk("ingress: src:%pi4 -> dst: %pi4", &ip->saddr, &ip->daddr);

	// Drop packet if in map
	__u8 *dummy = bpf_map_lookup_elem(&ips, &ip->saddr);
	if (dummy)
		return TC_ACT_SHOT;

	return TC_ACT_UNSPEC;
}
``` 

Build the gadget:

```bash 
$ sudo -E ig image build . -t mytcgadget
```

And run it:
```bash 
$ sudo -E ig run mytcgadget --iface enp5s0 
INFO[0000] Experimental features enabled                
INFO[0000] Running. Press Ctrl + C to finish
```

check trace_pipe to see the debug messages:
```bash
$ sudo cat /sys/kernel/debug/tracing/trace_pipe
          <idle>-0       [010] d.s.. 32963.221506: bpf_trace_printk: comm: wget, pid: 105845
          <idle>-0       [010] d.s.. 32963.227058: bpf_trace_printk: egress: src:192.178.050.068 -> dst: 172.017.000.004
          <idle>-0       [010] d.s.. 32963.227062: bpf_trace_printk: comm: wget, pid: 105845
          <idle>-0       [010] d.s.. 32963.227071: bpf_trace_printk: egress: src:172.017.000.004 -> dst: 192.178.050.068
          <idle>-0       [010] d.s.. 32963.227072: bpf_trace_printk: comm: wget, pid: 105845
          <idle>-0       [010] d.s.. 32963.227944: bpf_trace_printk: egress: src:192.178.050.068 -> dst: 172.017.000.004
          <idle>-0       [010] d.s.. 32963.227947: bpf_trace_printk: comm: wget, pid: 105845
          <idle>-0       [010] d.s.. 32963.227952: bpf_trace_printk: egress: src:172.017.000.004 -> dst: 192.178.050.068
          <idle>-0       [010] d.s.. 32963.227953: bpf_trace_printk: comm: wget, pid: 105845
```

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/2369

TODO:
- [X] How to handle multiple TC gadgets at the same time
- [X] How to handle multiple instances of ig at the same time 
- [X] How to handle collisions with other applications using TC programs (clsact qdisc and filters)
- [ ] Understand how the --iface parameter should behave when there are containers present
